refresh.template.py

```diff
@@ -147,6 +148,17 @@ def _get_headers_msvc(compile_args: typing.List[str], source_path: str):
         # End:   template filled by Bazel
     ))
 
+    # Write header_cmd to a temporary file, so we can use it as a parameter to cl.exe,
+    # because Windows cmd has a limitation of 8KB of command line length. So here we make a threshold of len(compile_args) < 8000
+    WIN_CMD_LIMITS = 8000
```

Comment:
Worth documenting the 8191--and why you chose a smaller limit? I assume because length calculate is an approximation of the real one with escaping https://docs.python.org/3/library/subprocess.html#converting-an-argument-sequence-to-a-string-on-windows

Another, (maybe better) approach could be try-except whatever exception it throws when the command is too long.

Reply:
Sure, will add better doc here

Reply:
S.g.  Thoughts on the pro/con of the try-except approach?

Reply:
try-except will be conflict with the following logic:
```python
for line in header_search_process.stderr.splitlines():
```
We can't detect whether the error comes from compile failure or command line overflow. Not sure if errorcode can help.

Reply:
Maybe I'm missing something, but I was thinking that there's probably a (Python) Error thrown if the line is longer than the command length limit (after escaping and all that). Like this one https://stackoverflow.com/questions/2381241/what-is-the-subprocess-popen-max-length-of-the-args-parameter?  You could test by exceeding the limit to subprocess.run and seeing what it throws!
(Let's definitely not rely on parsing an error message out of stderr--your hardcoded 8000 is more reliable than that!)

Reply:
Thanks for doing these, and experimenting around!

Reply:
@cpsauer Tested on my Windows 10 21H2:
For `powershell`, it can support user to type more than 8191 charactors on the console, while for `cmd`, user can't type any more if command line reaches 8191 charactors.
If you execute command from a script (i.e. `.ps1` or `.cmd/.bat`), the cmd may ignore charactors that exceeds 8191 limitation, while powershell doesn't. For example:

**CMD:**
```batch
set a=1234...(more than 8191 chars)....EOL
:: will only show 1234....(to 8191 length), can't echo "EOL", and no error
echo %a%
```

**PWSH:**
```powershell
$env:a="1234....(more than 8191 chars)...EOL"
# show full $a string, everything is OK
echo $env:a
```

But if your file path is too long (by default it can't longer than 254), most applications will fail with an error (just as your link) with both cmd and powershell console.

Reply:
Ah, sure, but I think we're talking past each other somehow.

I meant that *Python* would probably throw an Error if the args to subprocess.run were too long. That is, is an error like the one linked. Is such an error thrown if you temporarily replace `header_cmd` in the subprocess command with `"a "*10000` (or similar)?

Reply:
Oh, got your point.
Just test and `subprocess.call` will return error code **1**, with an stderr output "_command line too long_".

We may check return code **1** or stderr text (the text is in locale language)

Reply:
Refer to https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/1bc92ddf-b79e-413c-bbaa-99a5281a6c90
msvc compiler (only tested cl.exe) normally returns error code 2 (such as command error, compile error...)
We will use error code 1 for **command line too long** problem

Reply:
Oh, bummer. So no Python error, then. Sorry for pushing in the wrong direction.

Do whatever you think is safest and most robust here. If you're confident code 1 is always this command-length-issue, then do that. And if not, then let's fall back to your old solution. (Again, sorry.)